source/common/tcp_proxy/upstream.cc

```diff
@@ -526,7 +524,17 @@ void CombinedUpstream::onUpstreamTrailers(Http::ResponseTrailerMapPtr&& trailers
   responseDecoder().decodeTrailers(std::move(trailers));
 }
 
-Http::RequestHeaderMap* CombinedUpstream::downstreamHeaders() { return downstream_headers_.get(); }
+Http::RequestHeaderMap* CombinedUpstream::downstreamHeaders() {
+  if ((type_ != Http::CodecType::HTTP1) && (config_.usePost())) {
```

Comment:
Can we further move this setting logic to the `recordUpstreamSslConnection`? So that we only update the downstream_header once to save cost. This method can then just return the pointer that is used in a lot of other places.